### PR TITLE
webgpu: Update wgpu to 0.19

### PR DESCRIPTION
**Pull request by sagudev**
Instead of using 0.19, we use even newer version that removed IdentityFactories completely (also needed for #32008). I implemented proper drops for GPU* structs that are analogues to [Cleanup functions in firefox](https://searchfox.org/mozilla-central/rev/6121b33709dd80979a6806ff59096a561e348ae8/dom/webgpu/Adapter.cpp#206).

Relevant FF commit: https://github.com/mozilla/gecko-dev/commit/62af8e706eb426c60bf0f15db658dea1e507300a

try run: https://github.com/sagudev/servo/actions/runs/8820854175/job/24217313911 (crashes on `webgpu:shader,execution,expression,*` are #31397, timeouts are due too same problem as in #32008)

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes WebGPU CTS

### CODE REVIEW

**Comment by sagudev** [comment]
Weird crash on `_webgpu/webgpu/cts.https.html?q=webgpu:api,validation,queue,destroyed,texture:writeTexture:*`:

```
Texture[Id(1,1,vk)] does not exist (thread WGPU, at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-core-0.19.3/src/storage.rs:113)
2024-04-05T05:38:36.3419584Z          0: servoshell::backtrace::print
2024-04-05T05:38:36.3419696Z          1: servoshell::main::{{closure}}
2024-04-05T05:38:36.3419926Z          2: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
2024-04-05T05:38:36.3420387Z                    at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/boxed.rs:2021:9
2024-04-05T05:38:36.3420532Z             std::panicking::rust_panic_with_hook
2024-04-05T05:38:36.3420847Z                    at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs:735:13
2024-04-05T05:38:36.3421011Z          3: std::panicking::begin_panic_handler::{{closure}}
2024-04-05T05:38:36.3421326Z                    at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs:609:13
2024-04-05T05:38:36.3421505Z          4: std::sys_common::backtrace::__rust_end_short_backtrace
2024-04-05T05:38:36.3421862Z                    at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/sys_common/backtrace.rs:170:18
2024-04-05T05:38:36.3421964Z          5: rust_begin_unwind
2024-04-05T05:38:36.3422271Z                    at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs:597:5
2024-04-05T05:38:36.3422387Z          6: core::panicking::panic_fmt
2024-04-05T05:38:36.3422691Z                    at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/panicking.rs:72:14
2024-04-05T05:38:36.3423010Z          7: wgpu_core::storage::Storage<T,I>::get
2024-04-05T05:38:36.3423154Z          8: wgpu_core::registry::Registry<I,T>::get
2024-04-05T05:38:36.3423431Z          9: wgpu_core::device::queue::<impl wgpu_core::global::Global<G>>::queue_write_texture
2024-04-05T05:38:36.3423528Z         10: webgpu::WGPU::run
2024-04-05T05:38:36.3423720Z         11: std::sys_common::backtrace::__rust_begin_short_backtrace
2024-04-05T05:38:36.3423899Z         12: core::ops::function::FnOnce::call_once{{vtable.shim}}
2024-04-05T05:38:36.3424151Z         13: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
2024-04-05T05:38:36.3424452Z                    at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/boxed.rs:2007:9
2024-04-05T05:38:36.3424679Z             <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
2024-04-05T05:38:36.3424979Z                    at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/boxed.rs:2007:9
2024-04-05T05:38:36.3425139Z             std::sys::unix::thread::Thread::new::thread_start
2024-04-05T05:38:36.3425477Z                    at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/sys/unix/thread.rs:108:17
2024-04-05T05:38:36.3425563Z         14: <unknown>
2024-04-05T05:38:36.3425646Z         15: <unknown>
```

**Comment by sagudev** [comment]
Crash is from https://bugzilla.mozilla.org/show_bug.cgi?id=1860958

Need to recheck all drop/destroy.

**Comment by sagudev** [comment]
Unexpected results in first 7 chunks: [Chunks.txt](https://github.com/servo/servo/files/14971845/Chunks.txt). I will add `--update-good` in moz-webgpu-cts that will update only good results (PASS).

**Comment by sagudev** [comment]
Current results: https://github.com/sagudev/servo/actions/runs/8771474881/job/24069781625 a lot less crashes due to proper drops, one new crash https://gpuweb.github.io/cts/standalone/?q=webgpu:api,validation,createBindGroup:binding_resources,device_mismatch:entry={"texture":{"multisampled":false}} that is also reproducible on firefox https://crash-stats.mozilla.org/report/index/a13e187c-1065-4b5c-8a10-fee080240421

**Comment by sagudev** [comment]
Encoder timeouts are another example of https://github.com/servo/servo/pull/32008 (if poll is done on every iteration of loop everything works correctly). So I will update expectations as they are now, so we can get this moving.

**Comment by sagudev** [comment]
Useful patch when you want results of wpt run: https://github.com/servo/servo/pull/31995/commits/167137834c2306469831d5b3000180acc56b4177

**Comment by mrobinson** [comment]
> Useful patch when you want results of wpt run: [1671378](https://github.com/servo/servo/commit/167137834c2306469831d5b3000180acc56b4177)

Maybe we should disable the dashboard for all try runs.

**Review by gterzian — APPROVED** [review]
I'm OK with this short-term, but I think we need to re-design the lifecyle management of all these objects, because now there is too much back and forth over IPC and also I think we want to forbid `Drop` for dom objects, and also because I think by the time something drops it's not clear if we are shutting down or if it's just that objects that is dropped. 

This needs more work on the details, but I think we should do something like desribed in https://github.com/servo/servo/issues/32148

File: components/script/dom/identityhub.rs [hunk 1]
```diff
@@ -93,95 +129,97 @@ impl Identities {
...
     }
 
     pub fn kill_render_pipeline_id(&mut self, id: RenderPipelineId) {
         self.select(id.backend()).render_pipelines.free(id);
     }
 
     pub fn create_texture_id(&mut self, backend: Backend) -> TextureId {
-        self.select(backend).textures.alloc(backend)
+        let r = self.select(backend).textures.process(backend);
+        warn!("TEXTURE: {r:?}");
```
[the remark this answers is not in the document]
Comment (sagudev) [reply components/script/dom/identityhub.rs:205]: This should be removed as I used for debugging.

File: components/script/dom/gpuadapter.rs [hunk 2]
```diff
@@ -86,6 +86,21 @@ impl GPUAdapter {
     }
 }
 
+impl Drop for GPUAdapter {
```
[the remark this answers is not in the document]
Comment (sagudev) [reply components/script/dom/gpuadapter.rs:89]: I studied https://github.com/servo/servo/issues/26488 and it says:
> there was agreement that it should be possible to run some code in response to a GC finalizer (eg. webgl objects need to be able to unreference their graphics resources)

iiuc #26488 is about not accessing js managed types in Drop impl (as they can be already dropped). Also this is how it's done in gecko too.

File: components/webgpu/lib.rs [hunk 3]
```diff
@@ -1265,7 +1297,100 @@ impl<'a> WGPU<'a> {
...
+                        let global = &self.global;
+                        gfx_select!(id => global.render_bundle_drop(id));
+                        if let Err(e) = self.script_sender.send(WebGPUMsg::FreeRenderBundle(id)) {
+                            warn!("Unable to send FreeRenderBundle({:?}) ({:?})", id, e);
+                        };
+                    },
+                    WebGPURequest::DropQuerySet(id) => {
+                        let global = &self.global;
+                        gfx_select!(id => global.query_set_drop(id));
+                        if let Err(e) = self.script_sender.send(WebGPUMsg::FreeQuerySet(id)) {
```
[the remark this answers is not in the document]
Comment (sagudev) [reply components/webgpu/lib.rs:1391]: Actually, that's how it was done now (or it wasn't done at all). Factories had script_sender and send free commands to script thread. Let's discuss in https://github.com/servo/servo/issues/32148 what can be done for followup.

**Pull request merged**